test/DataAvailability.ts

```diff
@@ -0,0 +1,281 @@
+          .uploadData(timestamp, users, values2),
+      ).to.be.revertedWith('User already has data at this timestamp')
+
+      // Verify the user's last timestamp is still correct
+      expect(
+        await dataAvailability.getLastTimestamp(extraUser.address),
+      ).to.equal(timestamp)
+    })
+
+    it('should not allow non-uploader to upload data', async () => {
+      const timestamp = Math.floor(Date.now() / 1000)
+      const users = [extraUser.address]
+      const values = [100]
+
+      await expect(
+        dataAvailability
+          .connect(extraUser)
+          .uploadData(timestamp, users, values),
+      ).to.be.revertedWithCustomError(
+        dataAvailability,
+        'AccessControlUnauthorizedAccount',
+      )
+    })
+
+    it('should require timestamp to be greater than most recent', async () => {
+      const timestamp1 = Math.floor(Date.now() / 1000)
+      const timestamp2 = timestamp1 - 1
+      const users = [extraUser.address]
+      const values = [100]
+
+      await dataAvailability
+        .connect(uploader)
+        .uploadData(timestamp1, users, values)
+      await expect(
+        dataAvailability
+          .connect(uploader)
+          .uploadData(timestamp2, users, values),
+      ).to.be.revertedWith(
+        'Timestamp must be greater than or equal to most recent timestamp',
+      )
+    })
+
+    it('should calculate correct rolling hash', async () => {
+      const timestamp = Math.floor(Date.now() / 1000)
+      const users = [extraUser.address, uploader.address]
+      const values = [100, 200]
+      await dataAvailability
+        .connect(uploader)
+        .uploadData(timestamp, users, values)
+
+      // keccak256(encodePacked(['address', 'uint256'], ['0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC', 100]))
+      // > '0x0ff61c877bedc316874733e2574c5da6869d40d55e5cde4bcb053218179de1e3'
+      // keccak256(encodePacked(['address', 'uint256'], ['0x70997970C51812dc3A010C7d01b50e0d17dc79C8', 200]))
+      // > '0x2ed43a94c31997930a2c6fe97caf231eb915ef2b55b3477460f52c5648326b3b'
+      // XOR of the two:
+      // '0x21222613b8f454858d6b5c0b2be37eb83f88affe0bef993fabf01e4e5faf8ad8'
+
+      const expectedHash =
+        '0x21222613b8f454858d6b5c0b2be37eb83f88affe0bef993fabf01e4e5faf8ad8'
+      const hash = await dataAvailability.getHash(timestamp)
+      expect(hash).to.equal(expectedHash)
+    })
+    it('hash calculation should be commutative', async () => {
+      // Same as above, but in reverse order
+      const timestamp = Math.floor(Date.now() / 1000)
+      const users = [uploader.address, extraUser.address]
+      const values = [200, 100]
+      await dataAvailability
+        .connect(uploader)
+        .uploadData(timestamp, users, values)
+
+      const expectedHash =
+        '0x21222613b8f454858d6b5c0b2be37eb83f88affe0bef993fabf01e4e5faf8ad8'
+      const hash = await dataAvailability.getHash(timestamp)
+      expect(hash).to.equal(expectedHash)
+    })
```

Comment:
Test would probably be better if you just calculated each hash and asserted that they were equal instead of matching the `expectedHash`

Reply:
yeah fair enough, i'll update to actually perform the calc within the test itself